File: src/datastructures/singleylinkedlist.cpp
#include <iostream>

class SLL {

    struct Node {
        int data;
        Node* next;
    };

    public:
        void Insert(const int data);
        void Remove(const int data);
        void Print();

    private:
        Node *head = nullptr;
        Node *curr = nullptr;
};

///

void SLL::Insert(const int data) {
    Node* newNode = new Node();
    newNode->data = data;
    if(head == nullptr) head = newNode;
    else {
        curr = head;
        while(curr->next != nullptr) curr = curr->next;
        curr->next = newNode;
    }
}

void SLL::Remove(const int data) {
    Node* temp = nullptr;
    curr = head;
    while(curr != nullptr && curr->data != data) {
        temp = curr;
        curr = curr->next;
    } if(curr->data == head->data) {
        head = head->next;
    } else if(curr->data == data) {
        temp->next = curr->next;
    }
}

void SLL::Print() {
    curr = head;
    while(curr != nullptr) {
        std::cout << curr->data << " ";
        curr = curr->next;
    }
}

///

int main() {
    SLL list;
    list.Insert(1);
    list.Insert(2);
    list.Insert(3);
    list.Insert(4);
    list.Insert(5);
    list.Remove(3);
    list.Remove(1);
    list.Remove(5);
    list.Print();
}

File: src/datastructures/singleylinkedlist.java
class Node {
   
   public int  data;
   public Node next;
   
   public void DisplayNode() {
      System.out.println(data);
   }
}

public class SLL {
   
   private Node head = null;
   private Node current = null;
   
   public boolean isEmpty() {
      return (head == null);
   }
   
   public void InsertHead(int data) {
      Node newNode = new Node();
      newNode.data = data;
      head = newNode;
   }
   
   public void RemoveHead() {
      if(head != null) head = head.next;
   }
   
   public void InsertMiddle(int data) {
      Node newNode = new Node();
      newNode.data = data;
      Node current = head;
      while(current.next != null) current = current.next;
      current.next = newNode;
   }
   
   public void DeleteMiddle() {
      current = head;
      Node temp = null;
      while(current != head.next) {
         temp = current;
         current = current.next;
      } temp.next = current.next;
   }
   
   public void DeleteLast() {
      current = head;
      Node temp = null;
      while(current.next != null) {
         temp = current;
         current = current.next;
      } temp.next = current.next;  
   }
   
   public void Print() {
      current = head;
      while(current != null) {
         current.DisplayNode();
         current = current.next;
      }
   }
     
   public static void main(String args[]) {
      
      SLL test = new SLL();
      
      test.InsertHead(1);
      test.InsertMiddle(2);
      test.InsertMiddle(3);
      test.InsertMiddle(4);
      test.InsertMiddle(5);
      test.InsertMiddle(6);
      
      test.RemoveHead(); //remove 1
      test.DeleteMiddle(); //remove 3 because head is now 2
      test.DeleteLast(); //delete 6
      
      test.Print();
      
   }
}